proxy/http2/Http2Frame.cc

```diff
@@ -0,0 +1,253 @@
+/** @file
+
+  Http2Frame
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+#include "Http2Frame.h"
+
+//
+// Http2Frame
+//
+IOBufferReader *
+Http2Frame::reader() const
+{
+  return this->_ioreader;
+}
+
+const Http2FrameHeader &
+Http2Frame::header() const
+{
+  return this->_hdr;
+}
+
+bool
+Http2Frame::is_from_early_data() const
+{
+  return this->_from_early_data;
+}
+
+//
+// DATA Frame
+//
+int64_t
+Http2DataFrame::write_to(MIOBuffer *iobuffer) const
+{
+  // Write frame header
+  uint8_t buf[HTTP2_FRAME_HEADER_LEN];
+  http2_write_frame_header(this->_hdr, make_iovec(buf));
+  int64_t len = iobuffer->write(buf, sizeof(buf));
+
+  // Write frame payload
+  if (this->_reader && this->_payload_len > 0) {
+    int64_t written = 0;
+    // Fill current IOBufferBlock as much as possible to reduce SSL_write() calls
+    while (written < this->_payload_len) {
+      int64_t read_len = std::min(this->_payload_len - written, this->_reader->block_read_avail());
+      written += iobuffer->write(this->_reader->start(), read_len);
+      this->_reader->consume(read_len);
+    }
+    len += written;
```

Comment:
I know that we were playing with different versions of the Data frame writing.  One that wrote to contiguous buffers and passed that along to SSL_write().  Another that just passed along pointers from the original buffer (iobuffer in the case) and used the block pointers directly to write to SSL_write.  Did you find that performance was better with the extra intermediate copy to hopefully get bigger blocks?

Reply:
I compared these two approaches and the first approach is so much better.
The second one looks cool because of "no copy". However, it didn't improve performance as expected. 
`SSL_write()` is expensive than `memcpy()`, so reducing `SSL_write()` calls is key point, IMO.

Reply:
We can get rid of the `memcpy()` when SSL/TLS libraries provide lower APIs which slice `SSL_write()` functionalities.

Actually, our QUIC implementation keeps the IOBufferBlock chain to avoid `memcpy()`. Unfortunately, we can't take the same approach here, because it's using the EVP cipher functions directly.
https://github.com/apache/trafficserver/blob/504cc9fc924e9600f971d81231c5316b1c57e236/iocore/net/quic/QUICPacketPayloadProtector_openssl.cc#L30-L84